[AutoScaler V2] Invalid status transition from QUEUED to RAY_STOP_REQUESTED

### What happened + What you expected to happen

From Slack thread: https://ray.slack.com/archives/C02GFQ82JPM/p1764860436694129

I'm hitting an AssertionError with the v2 autoscaler.
It seems the Reconciler tries to terminate pending instances to enforce the limit, but the state machine fails because `QUEUED` -> `RAY_STOP_REQUESTED` is not a valid transition. There is some strange behaviour occurring because there should not be any scale up event to begin with.
The error:
```
AssertionError: Invalid status transition from QUEUED to RAY_STOP_REQUESTED
  File ".../ray/autoscaler/v2/instance_manager/instance_manager.py", line 264, in _update_instance
    assert InstanceUtil.set_status(instance, update.new_instance_status)
```

### Versions / Dependencies

ray: 2.49.1, 2.52.1

### Reproduction script

There is no reason for it to be removing nodes in the first place, it only happened when we increased our scale 10x

### Issue Severity

None

--- a/python/ray/autoscaler/v2/instance_manager/common.py
+++ b/python/ray/autoscaler/v2/instance_manager/common.py
@@ -200,6 +200,10 @@ class InstanceUtil:
                 # Cloud provider requested to launch a node for the instance.
                 # This happens when the a launch request is made to the node provider.
                 Instance.REQUESTED,
+                # Allocation request canceled before being requested.
+                # This happens when max_workers config is reduced or other termination
+                # triggers occur while the instance is still queued.
+                Instance.TERMINATED,
             },
             # When in this status, a launch request to the node provider is made.
             Instance.REQUESTED: {
@@ -333,10 +337,12 @@ class InstanceUtil:
                 # Retry the termination, become terminating again.
                 Instance.TERMINATING,
             },
-            # Whenever a cloud instance disappears from the list of running cloud
-            # instances from the node provider, the instance is marked as stopped. Since
-            # we guarantee 1:1 mapping of a Instance to a cloud instance, this is a
-            # terminal state.
+            # An instance is marked as terminated when:
+            # 1. A cloud instance disappears from the list of running cloud instances
+            #    from the node provider (follows from TERMINATING or other running states).
+            # 2. An allocation request is canceled before cloud resources are allocated
+            #    (follows from QUEUED).
+            # This is a terminal state.
             Instance.TERMINATED: set(),  # Terminal state.
             # When in this status, the cloud instance failed to be allocated by the
             # node provider.

--- a/python/ray/autoscaler/v2/instance_manager/reconciler.py
+++ b/python/ray/autoscaler/v2/instance_manager/reconciler.py
@@ -1177,7 +1177,16 @@ class Reconciler:
         # Add terminating instances.
         for terminate_request in to_terminate:
             instance_id = terminate_request.instance_id
-            if terminate_request.instance_status == IMInstance.ALLOCATED:
+            if terminate_request.instance_status == IMInstance.QUEUED:
+                # QUEUED instances have no cloud resources allocated yet.
+                # Cancel the allocation request by transitioning directly to TERMINATED.
+                updates[terminate_request.instance_id] = IMInstanceUpdateEvent(
+                    instance_id=instance_id,
+                    new_instance_status=IMInstance.TERMINATED,
+                    termination_request=terminate_request,
+                    details=f"allocation canceled: {terminate_request.details}",
+                )
+            elif terminate_request.instance_status == IMInstance.ALLOCATED:
                 # The instance is not yet running, so we can't request to stop/drain Ray.
                 # Therefore, we can skip the RAY_STOP_REQUESTED state and directly terminate the node.
                 im_instance_to_terminate = im_instances_by_instance_id[instance_id]

--- a/python/ray/autoscaler/v2/tests/test_instance_util.py
+++ b/python/ray/autoscaler/v2/tests/test_instance_util.py
@@ -34,7 +34,7 @@ class InstanceUtilTest(unittest.TestCase):
 
         g = InstanceUtil.get_valid_transitions()
 
-        assert g[Instance.QUEUED] == {Instance.REQUESTED}
+        assert g[Instance.QUEUED] == {Instance.REQUESTED, Instance.TERMINATED}
         all_status.remove(Instance.QUEUED)
 
         assert g[Instance.REQUESTED] == {

--- a/python/ray/autoscaler/v2/tests/test_reconciler.py
+++ b/python/ray/autoscaler/v2/tests/test_reconciler.py
@@ -1794,6 +1794,89 @@ class TestReconciler:
         assert instances["i-0"].status == Instance.TERMINATING
         assert instances["i-1"].status == Instance.TERMINATING
 
+    @staticmethod
+    def test_terminate_queued_instances_max_workers(setup):
+        instance_manager, instance_storage, _, cloud_resource_monitor = setup
+
+        instances = [
+            create_instance(
+                "head",
+                status=Instance.RAY_RUNNING,
+                node_kind=NodeKind.HEAD,
+                cloud_instance_id="c-head",
+                ray_node_id=binary_to_hex(b"r-head"),
+            ),
+            create_instance(
+                "i-0",
+                status=Instance.QUEUED,
+                instance_type="type-1",
+            ),
+            create_instance(
+                "i-1",
+                status=Instance.QUEUED,
+                instance_type="type-1",
+            ),
+        ]
+        TestReconciler._add_instances(instance_storage, instances)
+
+        # Empty list of Ray nodes - QUEUED instances have not launched ray yet
+        ray_nodes = []
+
+        # Cloud instance for head node only
+        cloud_instances = {
+            "c-head": CloudInstance("c-head", "head", True, NodeKind.HEAD),
+        }
+
+        # Scheduler should terminate QUEUED instances due to max_workers limit
+        mock_scheduler = MockScheduler(
+            to_launch=[],
+            to_terminate=[
+                TerminationRequest(
+                    id="t0",
+                    instance_id="i-0",
+                    instance_status=Instance.QUEUED,
+                    cause=TerminationRequest.Cause.MAX_NUM_NODE_PER_TYPE,
+                ),
+                TerminationRequest(
+                    id="t1",
+                    instance_id="i-1",
+                    instance_status=Instance.QUEUED,
+                    cause=TerminationRequest.Cause.MAX_NUM_NODE_PER_TYPE,
+                ),
+            ],
+        )
+
+        Reconciler.reconcile(
+            instance_manager=instance_manager,
+            scheduler=mock_scheduler,
+            cloud_provider=MagicMock(),
+            cloud_resource_monitor=cloud_resource_monitor,
+            ray_cluster_resource_state=ClusterResourceState(
+                node_states=ray_nodes,
+                cluster_resource_state_version=1,
+            ),
+            non_terminated_cloud_instances=cloud_instances,
+            cloud_provider_errors=[],
+            ray_install_errors=[],
+            autoscaling_config=MockAutoscalingConfig(
+                configs={
+                    "node_type_configs": {
+                        "type-1": {
+                            "name": "type-1",
+                            "resources": {"CPU": 1},
+                            "min_worker_nodes": 0,
+                            "max_worker_nodes": 0,
+                        }
+                    },
+                }
+            ),
+        )
+
+        instances, _ = instance_storage.get_instances()
+
+        assert instances["i-0"].status == Instance.TERMINATED
+        assert instances["i-1"].status == Instance.TERMINATED
+
 
 if __name__ == "__main__":
     if os.environ.get("PARALLEL_CI"):

--- a/src/ray/protobuf/instance_manager.proto
+++ b/src/ray/protobuf/instance_manager.proto
@@ -74,8 +74,9 @@ message Instance {
     // The instance is terminating - follows from the RAY_STOPPED state or ALLOCATED
     // state.
     TERMINATING = 9;
-    // The instance is terminated - follows from TERMINATING state or any other running
-    // states when instance was preempted.
+    // The instance is terminated - follows from TERMINATING state, QUEUED state when
+    // allocation request is canceled, or any other running states when instance was
+    // preempted.
     TERMINATED = 10;
     // Error states.
     // The instance allocation failed - follows from the REQUESTED state.

